learn/appendix_opengl/debug_callback.md

```diff
@@ -0,0 +1,158 @@
+
+    // Potentially, you may want to throw from the function for certain severity
+    // messages.
+    if (type == DebugType.DebugTypeError)
+    {
+        throw new Exception(message);
+    }
+}
+```
+
+> [!NOTE]
+> Neither the access modifier or whether the method is an instance method or a
+> static method matters.
+
+> [!WARNING]
+> This function is called from native code, which means there are stack frames
+> associated with them. If your debugger does not support mixed-mode debugging
+> it may affect your user experience if an exception is thrown or occurs within
+> the callback.
+
+> [!CAUTION]
+> The pointer `pMessage` is a a reference to memory you do not own. Attempts to
+> write to or accessing it outside the scope of the function can and will cause
+> an access violation. However, it is safe to use the string copied by the
+> `Marshal` class.
+
+Then you should create a delegate which encapsulates the function you just
+implmented.
+# [Base](#tab/delegate-gl)
+```cs
+private static GLDebugProc DebugMessageDelegate = OnDebugMessage;
+```
+# [KHR_debug](#tab/delegate-khr)
+```cs
+private static GLDebugProcKHR DebugMessageDelegate = OnDebugMessage;
+```
+# [ARB_debug_output](#tab/delegate-arb)
+```cs
+private static GLDebugProcARB DebugMessageDelegate = OnDebugMessage;
+```
+***
+<br/>
+
+> [!NOTE]
+> Creating this reference is critical. A delegate is managed by the .NET garbage
+> collector, however, this delegate will be passed to a native function as a
+> pointer. The garbage collector cannot track outside references, and if there
+> are no references to the delegate, the pointer passed to the native function
+> becomes "dangling" (in simpler terms, invalid). This causes an access violation
+> whenever OpenGL attempts to call you back.
+>
+> And as before, access modifiers and instance/static do not matter as long as it
+> suits your code.
+
+Finally you can provide OpenGL your delegate as your debug callback. You can now
+enable debug output, and optionally enable synchronous output.
+
+# [Base](#tab/enable-gl)
+```cs
+GL.DebugMessageCallback(DebugMessageDelegate, IntPtr.Zero);
+GL.Enable(EnableCap.DebugOutput);
+
+// Optionally
+GL.Enable(EnableCap.DebugOutputSynchronous)
+```
+# [KHR_debug](#tab/enable-khr)
+```cs
+GL.Khr.DebugMessageCallback(DebugMessageDelegate, IntPtr.Zero);
+GL.Enable(EnableCap.DebugOutput);
+
+// Optionally
+GL.Enable(EnableCap.DebugOutputSynchronous)
+```
+# [ARB_debug_output](#tab/enable-arb)
+```cs
+GL.Arb.DebugMessageCallback(DebugMessageDelegate, IntPtr.Zero);
+GL.Enable(EnableCap.DebugOutput);
+
+// Optionally
+GL.Enable(EnableCap.DebugOutputSynchronous)
+```
+***
+<br/>
+
+> [!TIP]
+> Using synchronous output may decrease your performance (untested) as OpenGL
+> will only call your function from the thread that the context is owned by.
+> Otherwise the graphics driver is allowed to call the function from any thread
+> concurrently. Be careful of the usual pitfalls of multithreading.
```

Comment:
I've tested synchronous vs no sync and there is a significant performance decrease with synchronous.

Enabling synchronous callbacks allows you to break into a debugger and look at the stacktrace for what function caused the issue, which is personally my most used use case for synchronous callbacks. So I think we should add a note about that.